packages/flutter/lib/src/material/material_state.dart

```diff
@@ -66,6 +68,163 @@ enum MaterialState {
+
+  /// Whether or not the child of this widget is considered to be
+  /// [MaterialState.pressed].
+  ///
+  /// If `null`, widget will be wrapped in a [Focus] to determine when the child
+  /// widget is in focus.
+  final bool pressed;
+
+  /// Whether or not the child of this widget is considered to be
+  /// [MaterialState.disabled].
+  ///
+  /// Defaults to `false`.
+  final bool disabled;
+
+  /// Whether or not the child of this widget is considered to be
+  /// [MaterialState.selected].
+  ///
+  /// Defaults to `false`.
+  final bool selected;
+
+  /// Whether or not the child of this widget is considered to be
+  /// [MaterialState.error].
+  ///
+  /// Defaults to `false`.
+  final bool error;
+
+  @override
+  _MaterialStateBuilderState createState() => _MaterialStateBuilderState();
+}
+
+class _MaterialStateBuilderState extends State<MaterialStateBuilder> {
+  final Set<MaterialState> _states = <MaterialState>{};
+
+  void _updateState(MaterialState state, bool value) {
+    if (_states.contains(state) != value) {
+      setState(() {
+        value ? _states.add(state) : _states.remove(state);
+      });
+    }
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    if (widget.hovered != null)
+      _updateState(MaterialState.hovered, widget.hovered);
+    if (widget.focused != null)
+      _updateState(MaterialState.focused, widget.focused);
+    if (widget.pressed != null)
+      _updateState(MaterialState.pressed, widget.pressed);
+    _updateState(MaterialState.disabled, widget.disabled);
+    _updateState(MaterialState.selected, widget.selected);
+    _updateState(MaterialState.error, widget.error);
+  }
+
+  @override
+  void didUpdateWidget(MaterialStateBuilder oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.hovered != null)
+      _updateState(MaterialState.hovered, widget.hovered);
+    if (widget.focused != null)
+      _updateState(MaterialState.focused, widget.focused);
+    if (widget.pressed != null)
+      _updateState(MaterialState.pressed, widget.pressed);
+    _updateState(MaterialState.disabled, widget.disabled);
+    _updateState(MaterialState.selected, widget.selected);
+    _updateState(MaterialState.error, widget.error);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    Widget result = widget.builder(context, _states);
+
+    if (widget.hovered == null) {
+      result = MouseRegion(
+        onEnter: (_) => _updateState(MaterialState.hovered, true),
+        onExit: (_) => _updateState(MaterialState.hovered, false),
+        child: result,
+      );
+    }
+
+    if (widget.focused == null) {
+      result = Focus(
+        skipTraversal: true,
+        onFocusChange: (bool value) => _updateState(MaterialState.focused, value),
+        child: result,
+      );
+    }
+
+    if (widget.pressed == null) {
+      result = GestureDetector(
+        excludeFromSemantics: true,
+        onForcePressStart: (_) => _updateState(MaterialState.pressed, true),
```

Comment:
Are the force press callbacks necessary?

Also as an FYI, Apple is deprecating force/3D touch and the latest iphones don't even have the hardware for it.

Reply:
Ah didn't know this, removed.